File: old-python/services/analytics.py
import json
import urllib.parse
from datetime import datetime, timedelta

from config.rules import get_food_subcategory, get_item_category
from database.connection import get_clean_receipts, get_db, get_state, set_state
from utils.formatters import format_food_stats, format_monthly_stats, format_weekly_report


def check_spending_anomaly(new_receipt_sum):
    with get_db() as conn:
        cursor = conn.cursor()
        cursor.execute("""
            SELECT SUM(total_sum)
            FROM receipts
            WHERE created_date >= date('now', '-30 days')
              AND created_date < date('now')
        """)
        row = cursor.fetchone()
        total_30days = row[0] if row and row[0] is not None else 0.0
        avg_daily = total_30days / 30.0

        cursor.execute("""
            SELECT SUM(total_sum)
            FROM receipts
            WHERE created_date >= date('now')
        """)
        row = cursor.fetchone()
        today_sum = (row[0] if row and row[0] is not None else 0.0) + new_receipt_sum

    if avg_daily > 200.0 and today_sum > 3 * avg_daily:
        today_str = datetime.now().date().isoformat()
        last_warning = get_state("last_anomaly_warning_date")
        if last_warning != today_str:
            set_state("last_anomaly_warning_date", today_str)
            ratio = today_sum / avg_daily
            return today_sum, avg_daily, ratio
    return None


def get_clean_items(start_date, end_date):
    receipts = get_clean_receipts(start_date, end_date)
    valid_items = []
    for _rkey, r in receipts.items():
        for it in r["items"]:
            if it["name"] in ["Платеж", "Предоплата", "Аванс"]:
                continue
            valid_items.append((r["owner"], it["name"], it["sum"], r.get("owner_phone")))
    return valid_items


def categorize_items(items_list):
    grouped = {}
    for owner, name, val, _owner_phone in items_list:
        matched = get_item_category(name, owner)
        grouped[matched] = grouped.get(matched, 0.0) + val
    return grouped


def build_weekly_report():
    now = datetime.now()
    today = now.date()
    start_of_this_week = today - timedelta(days=today.weekday())
    start_of_last_week = start_of_this_week - timedelta(days=7)

    end_of_last_week = start_of_this_week - timedelta(seconds=1)
    end_of_this_week = start_of_this_week + timedelta(days=7) - timedelta(seconds=1)

    dt_this_start = start_of_this_week.strftime("%Y-%m-%dT00:00:00")
    dt_this_end = end_of_this_week.strftime("%Y-%m-%dT23:59:59")
    dt_last_start = start_of_last_week.strftime("%Y-%m-%dT00:00:00")
    dt_last_end = end_of_last_week.strftime("%Y-%m-%dT23:59:59")

    items_this = get_clean_items(dt_this_start, dt_this_end)
    items_last = get_clean_items(dt_last_start, dt_last_end)

    cats_this = categorize_items(items_this)
    cats_last = categorize_items(items_last)

    total_this = sum(val for cat, val in cats_this.items() if cat != "Доходы")
    total_last = sum(val for cat, val in cats_last.items() if cat != "Доходы")

    items_this_clean = [(owner, name, val) for owner, name, val, owner_phone in items_this]

    html = format_weekly_report(
        start_of_this_week,
        today,
        start_of_last_week,
        end_of_last_week,
        cats_this,
        cats_last,
        total_this,
        total_last,
        items_this_clean,
    )
    return html


def build_food_report():
    now = datetime.now()
    start_of_month = now.strftime("%Y-%m-01T00:00:00")
    end_of_month = now.strftime("%Y-%m-%dT23:59:59")

    items = get_clean_items(start_of_month, end_of_month)
    food_expenses = {}
    total_food = 0.0

    for owner, name, val, _owner_phone in items:
        subcat = get_food_subcategory(name, owner)
        if subcat:
            food_expenses[subcat] = food_expenses.get(subcat, 0.0) + val
            total_food += val

    if total_food == 0:
        return None

    html = format_food_stats(total_food, food_expenses)
    return html


def build_monthly_stats_report():
    now = datetime.now()
    start_of_month = now.strftime("%Y-%m-01T00:00:00")
    end_of_month = now.strftime("%Y-%m-%dT23:59:59")

    items = get_clean_items(start_of_month, end_of_month)
    cats = categorize_items(items)
    total = sum(val for cat, val in cats.items() if cat != "Доходы")

    if total == 0 and cats.get("Доходы", 0.0) == 0:
        return None, None, 0.0

    total_lesha = sum(
        val
        for owner, name, val, owner_phone in items
        if owner_phone == "79639629392" and get_item_category(name, owner) != "Доходы"
    )
    total_masha = sum(
        val
        for owner, name, val, owner_phone in items
        if owner_phone == "79013652064" and get_item_category(name, owner) != "Доходы"
    )

    sorted_cats = sorted([(cat, val) for cat, val in cats.items() if cat != "Доходы"], key=lambda x: x[1], reverse=True)
    labels = []
    data = []
    for cat, val in sorted_cats:
        share = (val / total * 100) if total > 0 else 0.0
        labels.append(f"{cat} ({share:.1f}%)")
        data.append(round(val, 2))

    chart_config = {
        "type": "doughnut",
        "data": {
            "labels": labels,
            "datasets": [
                {"data": data, "backgroundColor": ["#5C6BC0", "#26A69A", "#FFA726", "#EC407A", "#AB47BC", "#78909C"]}
            ],
        },
        "options": {"title": {"display": True, "text": f"Расходы за текущий месяц (Итого: {total:.0f} ₽)"}},
    }

    config_str = json.dumps(chart_config)
    encoded_config = urllib.parse.quote(config_str)
    chart_url = f"https://quickchart.io/chart?c={encoded_config}&w=600&h=450"

    html = format_monthly_stats(total, cats, total_lesha=total_lesha, total_masha=total_masha)
    return chart_url, html, total


def build_taxi_report():
    now = datetime.now()
    start_of_month = now.strftime("%Y-%m-01")

    with get_db() as conn:
        cursor = conn.cursor()
        cursor.execute(
            """
            SELECT COUNT(*), SUM(total_cost), SUM(distance_km), SUM(duration_mins), SUM(tips_cost)
            FROM taxi_trips
            WHERE date >= ?
        """,
            (start_of_month,),
        )
        row = cursor.fetchone()

    if not row or row[0] == 0:
        return None, None

    count, total_cost, total_dist, total_dur, total_tips = row
    total_cost = total_cost or 0.0
    total_dist = total_dist or 0.0
    total_dur = total_dur or 0.0
    total_tips = total_tips or 0.0

    avg_price_km = total_cost / total_dist if total_dist > 0 else 0.0

    # Популярный тариф
    with get_db() as conn:
        cursor = conn.cursor()
        cursor.execute(
            """
            SELECT tariff_class, COUNT(*) as c
            FROM taxi_trips
            WHERE date >= ?
            GROUP BY tariff_class
            ORDER BY c DESC
            LIMIT 1
        """,
            (start_of_month,),
        )
        t_row = cursor.fetchone()
        popular_tariff = t_row[0] if t_row else "Неизвестно"

    # Топ-3 адресов назначения
    with get_db() as conn:
        cursor = conn.cursor()
        cursor.execute(
            """
            SELECT to_address, COUNT(*) as c
            FROM taxi_trips
            WHERE date >= ? AND to_address != 'Неизвестно'
            GROUP BY to_address
            ORDER BY c DESC
            LIMIT 3
        """,
            (start_of_month,),
        )
        addr_rows = cursor.fetchall()

    chart_url = None
    with get_db() as conn:
        cursor = conn.cursor()
        cursor.execute(
            """
            SELECT tariff_class, SUM(total_cost)
            FROM taxi_trips
            WHERE date >= ?
            GROUP BY tariff_class
        """,
            (start_of_month,),
        )
        tariff_costs = cursor.fetchall()

    if len(tariff_costs) > 1:
        labels = [f"{t} ({c:.0f} ₽)" for t, c in tariff_costs]
        data = [round(c, 2) for t, c in tariff_costs]
        chart_config = {
            "type": "doughnut",
            "data": {
                "labels": labels,
                "datasets": [{"data": data, "backgroundColor": ["#5C6BC0", "#26A69A", "#FFA726", "#EC407A"]}],
            },
            "options": {"title": {"display": True, "text": "Траты на такси по тарифам"}},
        }
        config_str = json.dumps(chart_config)
        encoded_config = urllib.parse.quote(config_str)
        chart_url = f"https://quickchart.io/chart?c={encoded_config}&w=400&h=300"

    month_ru = [
        "Январь",
        "Февраль",
        "Март",
        "Апрель",
        "Май",
        "Июнь",
        "Июль",
        "Август",
        "Сентябрь",
        "Октябрь",
        "Ноябрь",
        "Декабрь",
    ][now.month - 1]

    html = f"<h1>🚕 Статистика поездок: {month_ru} 2026</h1>"
    html += f"<p>💰 <b>Всего потрачено:</b> {total_cost:.2f} ₽ (Чаевые: {total_tips:.0f} ₽)</p>"
    html += f"<p>🛣 <b>Общий пробег:</b> {total_dist:.1f} км</p>"
    html += f"<p>⏱ <b>Время в пути:</b> {total_dur:.0f} мин</p>"
    html += f"<p>💳 <b>Средняя цена 1 км:</b> {avg_price_km:.2f} ₽/км</p>"
    html += f"<p>🚗 <b>Частый тариф:</b> {popular_tariff}</p>"

    if addr_rows:
        html += "<h3>📍 Популярные направления:</h3><ul>"
        for addr, count_addr in addr_rows:
            addr_short = addr.split(",")[0] if "," in addr else addr
            html += f"<li><b>{count_addr} раз(а)</b> — {addr_short}</li>"
        html += "</ul>"

    return chart_url, html

File: src/services/analytics.ts
import type { OpenDatabase } from "../db/client.js";
import { formatLocalDate, formatLocalDateTime, parseStoredDate, startOfLocalDay } from "../db/dates.js";
import { getCleanReceipts, getLatestReceipt, getState, setState } from "../db/operations.js";
import { filterDuplicateReceipts } from "./deduplicator.js";
import { formatFoodStats, formatMonthlyStats, formatWeeklyReport, monthName } from "./formatters.js";
import { getFoodSubcategory, getItemCategory } from "./rules.js";

export function getCleanItems(
  database: OpenDatabase,
  startDate: string,
  endDate: string,
): Array<{ owner: string | null; name: string; sum: number; ownerPhone: string | null }> {
  const receipts = getCleanReceipts(database, startDate, endDate);
  const ignored = filterDuplicateReceipts(receipts);
  const result: Array<{ owner: string | null; name: string; sum: number; ownerPhone: string | null }> = [];
  for (const [key, receipt] of receipts) {
    if (ignored.has(key)) continue;
    for (const item of receipt.items) {
      if (["Платеж", "Предоплата", "Аванс"].includes(item.name)) continue;
      result.push({ owner: receipt.owner, name: item.name, sum: item.sum, ownerPhone: receipt.ownerPhone });
    }
  }
  return result;
}

export function categorizeItems(
  items: Array<{ owner: string | null; name: string; sum: number }>,
): Record<string, number> {
  const result: Record<string, number> = {};
  for (const item of items) {
    const category = getItemCategory(item.name, item.owner);
    result[category] = (result[category] ?? 0) + item.sum;
  }
  return result;
}

export function sumExpenses(categories: Record<string, number>): number {
  return Object.entries(categories)
    .filter(([key]) => key !== "Доходы")
    .reduce((sum, [, value]) => sum + value, 0);
}

function totalExpenses(items: Array<{ owner: string | null; name: string; sum: number }>): number {
  return sumExpenses(categorizeItems(items));
}

export function checkSpendingAnomaly(database: OpenDatabase): [number, number, number] | undefined {
  const now = new Date();
  const todayStart = startOfLocalDay(now);
  const thirtyDaysAgo = new Date(todayStart.valueOf() - 30 * 86_400_000);
  const previousItems = getCleanItems(
    database,
    formatLocalDateTime(thirtyDaysAgo),
    formatLocalDateTime(new Date(todayStart.valueOf() - 1_000)),
  );
  const todayItems = getCleanItems(database, formatLocalDateTime(todayStart), formatLocalDateTime(now));
  const average = totalExpenses(previousItems) / 30;
  const todayTotal = totalExpenses(todayItems);
  if (average <= 200 || todayTotal <= 3 * average) return undefined;

  const today = formatLocalDate(now);
  if (getState(database, "last_anomaly_warning_date") === today) return undefined;
  setState(database, "last_anomaly_warning_date", today);
  return [todayTotal, average, todayTotal / average];
}

function monday(date: Date): Date {
  const result = new Date(date);
  result.setHours(0, 0, 0, 0);
  result.setDate(result.getDate() - result.getDay() + (result.getDay() === 0 ? -6 : 1));
  return result;
}

export function buildWeeklyReport(database: OpenDatabase): string {
  const today = new Date();
  const thisStart = monday(today);
  const lastStart = new Date(thisStart);
  lastStart.setDate(lastStart.getDate() - 7);
  const lastEnd = new Date(thisStart);
  lastEnd.setMilliseconds(-1);
  const thisEnd = new Date(thisStart);
  thisEnd.setDate(thisEnd.getDate() + 6);
  const thisItems = getCleanItems(database, formatLocalDateTime(thisStart), formatLocalDateTime(thisEnd));
  const lastItems = getCleanItems(database, formatLocalDateTime(lastStart), formatLocalDateTime(lastEnd));
  const thisCategories = categorizeItems(thisItems);
  const lastCategories = categorizeItems(lastItems);
  const totalThis = sumExpenses(thisCategories);
  const totalLast = sumExpenses(lastCategories);
  return formatWeeklyReport(
    thisStart,
    today,
    lastStart,
    lastEnd,
    thisCategories,
    lastCategories,
    totalThis,
    totalLast,
    thisItems,
  );
}

export function buildFoodReport(database: OpenDatabase): string | undefined {
  const now = new Date();
  const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
  const items = getCleanItems(database, formatLocalDateTime(monthStart), formatLocalDateTime(now));
  const expenses: Record<string, number> = {};
  let total = 0;
  for (const item of items) {
    const subcategory = getFoodSubcategory(item.name, item.owner ?? "");
    if (!subcategory) continue;
    expenses[subcategory] = (expenses[subcategory] ?? 0) + item.sum;
    total += item.sum;
  }
  return total === 0 ? undefined : formatFoodStats(total, expenses);
}

export function buildMonthlyStatsReport(database: OpenDatabase): { chartUrl?: string; html?: string; total: number } {
  const now = new Date();
  const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
  const items = getCleanItems(database, formatLocalDateTime(monthStart), formatLocalDateTime(now));
  const categories = categorizeItems(items);
  const total = sumExpenses(categories);
  if (total === 0 && (categories.Доходы ?? 0) === 0) return { total: 0 };

  const totalLesha = items
    .filter((item) => item.ownerPhone === "79639629392" && getItemCategory(item.name, item.owner) !== "Доходы")
    .reduce((sum, item) => sum + item.sum, 0);
  const totalMasha = items
    .filter((item) => item.ownerPhone === "79013652064" && getItemCategory(item.name, item.owner) !== "Доходы")
    .reduce((sum, item) => sum + item.sum, 0);
  const sorted = Object.entries(categories)
    .filter(([key]) => key !== "Доходы")
    .sort(([, a], [, b]) => b - a);
  const chart = {
    type: "doughnut",
    data: {
      labels: sorted.map(
        ([category, value]) => `${category} (${total > 0 ? ((value / total) * 100).toFixed(1) : "0.0"}%)`,
      ),
      datasets: [
        {
          data: sorted.map(([, value]) => Number(value.toFixed(2))),
          backgroundColor: ["#5C6BC0", "#26A69A", "#FFA726", "#EC407A", "#AB47BC", "#78909C"],
        },
      ],
    },
    options: { title: { display: true, text: `Расходы за текущий месяц (Итого: ${total.toFixed(0)} ₽)` } },
  };
  return {
    chartUrl: `https://quickchart.io/chart?c=${encodeURIComponent(JSON.stringify(chart))}&w=600&h=450`,
    html: formatMonthlyStats(total, categories, totalLesha, totalMasha),
    total,
  };
}

export function buildTaxiReport(database: OpenDatabase): { chartUrl?: string; html?: string } {
  const monthStart = new Date();
  monthStart.setDate(1);
  const start = formatLocalDate(monthStart);
  const summary = database.sqlite
    .query(
      "SELECT COUNT(*) AS count, SUM(total_cost) AS total_cost, SUM(distance_km) AS total_dist, SUM(duration_mins) AS total_dur, SUM(tips_cost) AS total_tips FROM taxi_trips WHERE date >= ?",
    )
    .get(start) as { count: number; total_cost?: number; total_dist?: number; total_dur?: number; total_tips?: number };
  if (!summary || summary.count === 0) return {};
  const popular = database.sqlite
    .query("SELECT tariff_class FROM taxi_trips WHERE date >= ? GROUP BY tariff_class ORDER BY COUNT(*) DESC LIMIT 1")
    .get(start) as { tariff_class?: string } | null;
  const addresses = database.sqlite
    .query(
      "SELECT to_address, COUNT(*) AS count FROM taxi_trips WHERE date >= ? AND to_address != 'Неизвестно' GROUP BY to_address ORDER BY count DESC LIMIT 3",
    )
    .all(start) as Array<{ to_address?: string; count: number }>;
  const tariffCosts = database.sqlite
    .query("SELECT tariff_class, SUM(total_cost) AS total FROM taxi_trips WHERE date >= ? GROUP BY tariff_class")
    .all(start) as Array<{ tariff_class?: string; total: number }>;
  const total = summary.total_cost ?? 0;
  const distance = summary.total_dist ?? 0;
  let html = `<h1>🚕 Статистика поездок: ${monthName(new Date().getMonth())} ${new Date().getFullYear()}</h1><p>💰 <b>Всего потрачено:</b> ${total.toFixed(2)} ₽ (Чаевые: ${(summary.total_tips ?? 0).toFixed(0)} ₽)</p><p>🛣 <b>Общий пробег:</b> ${distance.toFixed(1)} км</p><p>⏱ <b>Время в пути:</b> ${(summary.total_dur ?? 0).toFixed(0)} мин</p><p>💳 <b>Средняя цена 1 км:</b> ${(distance > 0 ? total / distance : 0).toFixed(2)} ₽/км</p><p>🚗 <b>Частый тариф:</b> ${popular?.tariff_class ?? "Неизвестно"}</p>`;
  if (addresses.length > 0) {
    html += "<h3>📍 Популярные направления:</h3><ul>";
    for (const address of addresses)
      html += `<li><b>${address.count} раз(а)</b> — ${(address.to_address ?? "").split(",")[0]}</li>`;
    html += "</ul>";
  }
  let chartUrl: string | undefined;
  if (tariffCosts.length > 1) {
    const chart = {
      type: "doughnut",
      data: {
        labels: tariffCosts.map((row) => `${row.tariff_class} (${row.total.toFixed(0)} ₽)`),
        datasets: [
          {
            data: tariffCosts.map((row) => Number(row.total.toFixed(2))),
            backgroundColor: ["#5C6BC0", "#26A69A", "#FFA726", "#EC407A"],
          },
        ],
      },
      options: { title: { display: true, text: "Траты на такси по тарифам" } },
    };
    chartUrl = `https://quickchart.io/chart?c=${encodeURIComponent(JSON.stringify(chart))}&w=400&h=300`;
  }
  return chartUrl ? { chartUrl, html } : { html };
}

export function getLatestDashboardReceipt(database: OpenDatabase): string {
  const latest = getLatestReceipt(database);
  return latest
    ? `🛍 ${parseStoredDate(latest.date).toLocaleString("ru-RU", { day: "2-digit", month: "2-digit", hour: "2-digit", minute: "2-digit" })} · ${latest.owner ?? "Неизвестный магазин"} · <b>${latest.total.toFixed(0)} ₽</b>`
    : "Нет трат";
}
